presto-product-tests/src/main/java/com/facebook/presto/tests/SqlCancelTests.java

```diff
@@ -0,0 +1,150 @@
+
+public class SqlCancelTests
+        extends ProductTest
+{
+    private ExecutorService executor;
+    private QueryCanceller queryCanceller;
+
+    @Inject
+    @Named("databases.presto.server_address")
+    private String serverAddress;
+
+    @BeforeTestWithContext
+    public void setUp()
+    {
+        executor = newFixedThreadPool(2);
+        queryCanceller = new QueryCanceller(serverAddress);
+    }
+
+    @AfterTestWithContext
+    public void cleanUp()
+    {
+        if (executor != null) {
+            executor.shutdownNow();
+        }
+        if (queryCanceller != null) {
+            queryCanceller.close();
+        }
+    }
+
+    @Test(groups = {CANCEL_QUERY}, timeOut = 60 * 1000L)
+    public void cancelCreateTableTest()
+            throws Exception
+    {
+        String tableName = format("cancel_createtable_%s", nanoTime());
+        String sql = format("CREATE TABLE %s AS SELECT * FROM tpch.sf1.lineitem", tableName);
+
+        runAndThenCancelQuery(sql);
+        assertThat(() -> query(format("SELECT * from %s", tableName)))
+                .failsWithMessage(format("Table hive.default.%s does not exist", tableName));
+    }
+
+    @Test(groups = {CANCEL_QUERY}, timeOut = 60 * 1000L)
+    public void cancelInsertIntoTest()
+            throws Exception
+    {
+        String tableName = format("cancel_insertinto_%s", nanoTime());
+        query(format("CREATE TABLE %s " +
+                "(orderkey BIGINT, partkey BIGINT, shipinstruct VARCHAR(25)) ", tableName));
+        String sql = format("INSERT INTO %s " +
+                "SELECT orderkey, partkey, shipinstruct FROM tpch.sf1.lineitem", tableName);
+
+        runAndThenCancelQuery(sql);
+        assertThat(query(format("SELECT * from %s", tableName))).hasNoRows();
+        query(format("DROP TABLE %s", tableName));
+    }
+
+    @Test(groups = {CANCEL_QUERY}, timeOut = 60 * 1000L)
+    public void cancelSelectTest()
+            throws Exception
+    {
+        runAndThenCancelQuery(format("SELECT * FROM tpch.sf1.lineitem AS cancel_select_%s", nanoTime()));
+    }
+
+    private void runAndThenCancelQuery(String sql)
+            throws Exception
+    {
+        Future<?> executeQueryThread = executor.submit(() -> assertThat(() -> query(sql)).failsWithMessage("Query was canceled"));
+        Future<Response> cancelQueryThread = executor.submit(() -> executeCancelable(sql));
+
+        try {
+            executeQueryThread.get(30, TimeUnit.SECONDS);
+        }
+        catch (TimeoutException e) {
+            executeQueryThread.cancel(true);
+            cancelQueryThread.cancel(true);
+            throw new RuntimeException("Query execution time limit exceeded", e);
+        }
+
+        Response response = cancelQueryThread.get();
+        if (response.getStatusCode() != HttpStatus.OK.code()
+                && response.getStatusCode() != HttpStatus.ACCEPTED.code()
+                && response.getStatusCode() != HttpStatus.NO_CONTENT.code()) {
```

Comment:
Can we have exactly one expected value for `response.getStatusCode()`?
Use `assertEquals` for comparing actual and expected

Reply:
I do not know. I would need to test it a lot to make sure as there might be some race conditions. Lets leave it as it is.